ENH: np.unique: support hash based unique for string dtype

math-hiyoko opened the pull request:
## Description
This PR introduces hash-based uniqueness extraction support for NPY_STRING, NPY_UNICODE, and NPY_VSTRING types in NumPy's np.unique function.  
The existing hash-based unique implementation, previously limited to integer data types, has been generalized to accommodate additional data types including string-related ones. Minor refactoring was also performed to improve maintainability and readability.  

## Benchmark Results
The following benchmark demonstrates significant performance improvement from the new implementation.   
The test scenario (1 billion strings array) follows the experimental setup described https://github.com/numpy/numpy/pull/26018#issuecomment-2012427568  
```PYTHON
import random
import string
import time

import numpy as np
import polars as pl

chars = string.ascii_letters + string.digits
arr = np.array(
    [
        ''.join(random.choices(chars, k=random.randint(5, 10)))
        for _ in range(1_000)
    ] * 1_000_000,
    dtype='T',
)
np.random.shuffle(arr)

time_start = time.perf_counter()
print("unique count (hash based): ", len(np.unique(arr)))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (polars): ", len(pl.Series(arr).unique()))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))
```

## Result
```SHELL
unique count (hash based):  1000
33.583 secs
unique count (numpy main):  1000
498.011 secs
unique count (polars):  1000
74.023 secs
```

close #28364

math-hiyoko committed:
Support NPY_STRING, NPY_UNICODE

math-hiyoko committed:
unique for NPY_STRING and NPY_UNICODE

math-hiyoko committed:
fix construct array

math-hiyoko committed:
remove unneccessary include

math-hiyoko committed:
refactor

math-hiyoko committed:
refactoring

math-hiyoko committed:
comment

math-hiyoko committed:
feature: unique for NPY_VSTRING

math-hiyoko committed:
refactoring

math-hiyoko committed:
remove unneccessary include

math-hiyoko committed:
add test

math-hiyoko committed:
add error message

math-hiyoko committed:
linter

math-hiyoko committed:
linter

math-hiyoko committed:
reserve bucket

math-hiyoko committed:
remove emoji from testcase

math-hiyoko committed:
fix testcase

math-hiyoko committed:
remove error

math-hiyoko committed:
fix testcase

math-hiyoko committed:
fix testcase name

math-hiyoko committed:
use basic_string

math-hiyoko committed:
fix testcase

math-hiyoko committed:
add ValueError

math-hiyoko committed:
fix testcase

math-hiyoko committed:
fix memory error

math-hiyoko committed:
remove multibyte char

math-hiyoko committed:
refactoring

math-hiyoko committed:
add multibyte char

math-hiyoko committed:
refactoring

math-hiyoko committed:
fix memory error

math-hiyoko committed:
fix GIL

math-hiyoko committed:
fix strlen

math-hiyoko committed:
remove PyArray_GETPTR1

math-hiyoko committed:
refactoring

math-hiyoko committed:
refactoring

math-hiyoko committed:
use optional

math-hiyoko committed:
refactoring

math-hiyoko committed:
refactoring

math-hiyoko committed:
refactoring

math-hiyoko committed:
refactoring

math-hiyoko committed:
fix comment

math-hiyoko committed:
linter

math-hiyoko committed:
add doc

math-hiyoko committed:
DOC: fix

math-hiyoko committed:
DOC: fix format

math-hiyoko committed:
MNT: refactoring

math-hiyoko committed:
MNT: refactoring

math-hiyoko commented:
Everything is ready now!!

seberg commented:
Thanks for looking at this! Ping @adrinjalali in case you have time/interest thinking about this.
I need to dive into this deeper, though.  So below some thoughts only.

For string/bytes (old style), it would seem reasonable to store a reference/pointer into the map and template the hash/equal function of the map appropriately (can also be done with a custom type as you did though).
For `StringDType` we also could store the full packed string, but operate on it again in hash/equal with the existing dtype/allocator.  However, at the end we will want to do a full copy (to not keep referencing the old arrays strings).

I.e. I am curious if we can do the map more light-weight, both because i wonder if it may be simple, but also because copying to `std::string` is maybe slower and possibly wrong (embedded `\0` characters).
It may well be that this thought only becomes useful if we refactor things first, though.

math-hiyoko commented:
Thanks for your comments!

I agree that storing just pointers (or references) in the hash set sounds like a good idea—I’ll give that approach a try.

I also realized I hadn’t fully considered the cases where strings include embedded \0 characters. Since handling embedded \0 correctly in C++‘s std::string can be tricky, this is another strong reason to handle strings as pointers directly. I’ll explore this further and update accordingly.

math-hiyoko committed:
ENH: Store pointers to strings in the set instead of the strings themselves.

math-hiyoko committed:
FIX: length in memcmp

math-hiyoko committed:
ENH: refactoring

math-hiyoko committed:
DOC: 49sec -> 34sec

math-hiyoko commented:
I’ve updated the implementation to store pointers in the unordered_set.

In the original PR, I attempted to unify the implementations using a single templated unique function. However, considering the complexity of the code, I’ve split this into three separate functions: unique_integer, unique_string, and unique_vstring.

This change has improved the execution speed from the original PR’s 49 seconds down to 34 seconds!!

ngoldbaum commented:
Wow, cool!

Totally missed this until now. Will try to give it a careful look.

ngoldbaum commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -361,6 +361,11 @@ def _unique1d(ar, return_index=False, return_inverse=False,
 
+    if ar.dtype.kind == 'T' and not equal_nan:
+        raise ValueError(
+            "Currently, `equal_nan` can only be True dtype is `T` (StringDType)."
```
```suggestion
            "Currently, `equal_nan` can only be True if dtype is `T` (StringDType)."
```

ngoldbaum commented on doc/release/upcoming_changes/28767.change.rst:
```diff
@@ -0,0 +1,9 @@
+As a result, calling np.unique() on an array of strings will use
+the faster hash-based method to obtain unique values.
+This also works for StringDType arrays containing None (missing values)
+when using equal_nan=True (treating missing values as equal).
```
I’m confused - this doesn’t say anything about why `unique_values` would return unsorted strings now.

math-hiyoko replied:
I’ve mentioned that the hash-based approach doesn’t guarantee sorted output.
The title of this doc follows the convention used in https://github.com/numpy/numpy/pull/26018

ngoldbaum reviewed (commented):
I still need to look more carefully at the C++ code but at a high level this looks good.

math-hiyoko committed:
Update numpy/lib/_arraysetops_impl.py

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

math-hiyoko committed:
DOC: Mention that hash-based np.unique returns unsorted strings

math-hiyoko committed:
Merge branch 'feature/#28364' of github.com:math-hiyoko/numpy into feature/#28364

ngoldbaum commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -836,6 +839,46 @@ class Subclass(np.ndarray):
+    def test_unique_vstring_hash_based(self):
+        # test for unicode and nullable string arrays
+        a = np.array(['straße', None, 'strasse', 'straße', None, 'niño', 'nino', 'élève', 'eleve', 'niño', 'élève'], dtype=StringDType(na_object=None))
+        unq_sorted_wo_none = ['eleve', 'nino', 'niño', 'strasse', 'straße', 'élève']
```
These strings should probably include items that exercise more of the functionality in StringDType internals. See NEP 55 and the StringDType tests for examples, but basically you only include "short" strings here, but probably also want to include both "medium" and "long" strings that are allocated on the heap outside of the ndarray buffer.

math-hiyoko replied:
I actually encountered memory errors when adding medium and long strings to the tests, so I’ve implemented fixes addressing those issues as well. Thanks for pointing that out!

math-hiyoko committed:
ENH: support medium and long vstrings

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +19,229 @@ struct FinalAction {
+// function to caluculate the hash of a string
+template <typename T>
+size_t str_hash(const T *str, npy_intp num_chars) {
+    // https://www.boost.org/doc/libs/1_88_0/libs/container_hash/doc/html/hash.html#notes_hash_combine
```
This link says boost hasn't used this algorithm since Boost 1.56 in 2014 and currently uses an algorithm based on murmurhash3.

Did you try FNV hash or another non-cryptographic hash? 

http://www.isthe.com/chongo/tech/comp/fnv/index.html

math-hiyoko replied:
I've updated the hash algorithm from the original implementation to FNV-1a.
Although the overall execution speed has slightly decreased due to other changes introduced simultaneously, the hash algorithm change itself resulted in a 94.4% reduction in hashing time.

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +19,229 @@ struct FinalAction {
+    npy_string_allocator *in_allocator = NpyString_acquire_allocator((PyArray_StringDTypeObject *)descr);
+    auto in_allocator_dealloc = finally([&]() {
+        NpyString_release_allocator(in_allocator);
+    });
```
nice, might be worth adding C++ wrappers for the `NpyString` API somewhere in NumPy so this pattern can be used elsewhere.

ngoldbaum commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -836,6 +839,104 @@ class Subclass(np.ndarray):
+        assert_equal(count_none, 1)
+
+        a1_wo_none = sorted(x for x in a1 if x is not None)
+        assert_array_equal(a1_wo_none, unq_sorted_wo_none)
```
Nice, I guess this handles missing strings fine.

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -100,46 +253,62 @@ unique(PyArrayObject *self)
+        if ((*it)->buf == NULL) {
+            NpyString_pack_null(out_allocator, packed_string);
+        } else {
+            NpyString_pack(out_allocator, packed_string, (*it)->buf, (*it)->size);
```
Both of these functions can fail and you have to check for and handle the error case.

math-hiyoko replied:
I’ve updated it so that if packing fails, it returns NULL.

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +19,229 @@ struct FinalAction {
+    npy_intp isize = PyArray_SIZE(self);
+
+    // Reserve hashset capacity in advance to minimize reallocations.
+    std::unordered_set<T> hashset(isize * 2);
```
Why `isize * 2`? Shouldn't `isize` (all distinct values) be enough?

math-hiyoko replied:
Actually, I set the hashset capacity to 2 * isize not just to avoid reallocations, but also to reduce hash collisions as much as possible. I've added a comment in the code to clarify this reasoning as well.
https://www.cse.cuhk.edu.hk/irwin.king/_media/teaching/csci2100b/csci2100b-2013-05-hash.pdf

seberg replied:
Relatively random comment, but I don't think we should reserve "enough" from the start (the `* 2` also looks a bit like it includes the hashmap load factor?).  This scheme is potentially larger than the original array, even though often the array may have very few unique elements.

If reserving is worth it, maybe it would make sense reserve a relatively large default with `min(isize, not_too_large)`, where `not_to_large` is maybe a few kiB?

math-hiyoko replied:
Since we’re now only storing pointers in the unordered_set, its size shouldn’t be larger than the original array—especially for strings, where the set just holds pointers rather than the string data itself. I agree that for integers, your suggested approach would be more appropriate.

seberg replied:
If strings are short that argument doesn't quite hold.  So for simplicity, I would suggest to just use the same scheme for everything.

math-hiyoko replied:
Got it. I’ll use `min(isize, not_too_large)` instead.
For not_too_large, I’m thinking of setting it to 1024 (since each element is a pointer, so that’s about 4 KiB).

math-hiyoko replied:
I’ve updated the code to use `min(isize, 1024)` as the initial bucket size for the hash set.  I’ve also added a comment in the code to explain this choice.

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +19,229 @@ struct FinalAction {
-    npy_intp *strideptr = NpyIter_GetInnerStrideArray(iter);
-    npy_intp *innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+    NPY_DISABLE_C_API;
+    PyThreadState *_save2 = PyEval_SaveThread();
```
you could use an RAII wrapper for this too.

math-hiyoko replied:
I’ve added an RAII wrapper for this now.

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +19,229 @@ struct FinalAction {
+    });
+    auto hash = [](const npy_static_string *value) -> size_t {
+        if (value->buf == NULL) {
+            return 0;
```
Is it a problem that missing strings all hash the same? Probably not?

math-hiyoko replied:
I actually think they should have the same hash value—otherwise, it wouldn’t be possible to implement behavior like equal_nan=True.

seberg replied:
It would be a problem if they are not considered equal.  In that case, this should hash based on the address.

math-hiyoko replied:
In that case, I’ll update the unique_vstring function to accept an equal_nan option and handle the behavior accordingly.

ngoldbaum reviewed (commented):
Left a few comments inline.

math-hiyoko committed:
FIX: comment

math-hiyoko committed:
ENH: use RAII wrapper

math-hiyoko committed:
FIX: error handling of string packing

math-hiyoko committed:
FIX: error handling of string packing

math-hiyoko committed:
FIX: change default bucket size

math-hiyoko committed:
FIX: include

math-hiyoko committed:
FIX: cast

math-hiyoko committed:
ENH: support equal_nan=False

math-hiyoko committed:
FIX: function equal

math-hiyoko committed:
FIX: check the case if pack_status douesn't return NULL

math-hiyoko committed:
FIX: check the case if pack_status douesn't return NULL

math-hiyoko committed:
FIX: stderr

seberg commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -4572,7 +4572,7 @@ static struct PyMethodDef array_module_methods[] = {
         METH_FASTCALL | METH_KEYWORDS, NULL},
     {"_unique_hash",  (PyCFunction)array__unique_hash,
-        METH_O, "Collect unique values via a hash map."},
+        METH_VARARGS | METH_KEYWORDS, "Collect unique values via a hash map."},
```
Please use `METH_FASTCALL | METH_KEYWORDS`, you'll have to change the parsing, but you should find an example easily (and the call overhead will be much smaller).

math-hiyoko committed:
ENH: METH_VARARGS -> METH_FASTCALL

math-hiyoko committed:
FIX: log

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +24,257 @@ struct FinalAction {
+    // NumPy API calls and Python object manipulations require holding the GIL.
+    Py_INCREF(descr);
+
+    // variables for the vstring, this operation require holding the GIL.
```
Why did you make this change? I don’t think this is true.

ngoldbaum replied:
The docstring for that function doesn’t say you need to hold the GIL to call it. It says you shouldn’t call into functions requiring the GIL, while holding the allocator lock, as that might deadlock.

Does that make sense? Is there a better way to phrase the docstring to avoid confusion?

math-hiyoko replied:
I understand now. I don't think the docstring is misleading.

math-hiyoko committed:
FIX: release allocator

math-hiyoko committed:
FIX: comment

math-hiyoko committed:
FIX: delete log

math-hiyoko committed:
ENH: implemented FNV-1a as hash function

math-hiyoko committed:
bool -> npy_bool

math-hiyoko committed:
FIX: cast

math-hiyoko committed:
34sec -> 35.1sec

math-hiyoko commented:
@seberg @ngoldbaum 
I've addressed all comments received so far.

math-hiyoko committed:
Merge branch 'main' into feature/#28364

math-hiyoko committed:
fix: lint

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +23,263 @@ struct FinalAction {
-template<typename T>
+// function to caluculate the hash of a string
+template <typename T>
+size_t str_hash(const T *str, npy_intp num_chars) {
```
Note that `npy_ucs4` is four bytes and fnv-1a operates on octets of data (e.g. individual bytes).

[The reference implementation](https://github.com/lcn2/fnv/blob/b7fcbee95538ee6a15744e756e7e7f1c02862cb0/hash_64a.c#L114-L116) takes a `void *` pointer and immediately casts it to `unsigned char *`. You could do similar.

We could also add the reference implementation as a vendored dependency (e.g. a git submodule). The license is compatible.

math-hiyoko replied:
Thanks for the suggestion.
If we go the submodule route for lcn2/fnv, where in the NumPy tree would you like the submodule to live? Let me know your preferred path and I’ll add it accordingly.

math-hiyoko replied:
Tried the submodule approach, but ran into two blockers:

- the reference source expects to be built with a plain `make` step that is not available on every platform/toolchain we target (Pyodide, MSVC, …)
- the code still uses legacy BSD typedefs such as `u_int32_t`; that builds fine on Linux/macOS but fails to compile on WASM, Windows/Clang, etc.

Because of these two issues the submodule route looks impractical for NumPy at the moment.

ngoldbaum replied:
Fair enough, thanks for checking. Vendoring a somewhat adapted version makes sense.

math-hiyoko replied:
I’ve re-implemented the FNV-1a hash function based on the reference implementation from [lcn2/fnv](https://github.com/lcn2/fnv), with appropriate modifications. To keep the scope and visibility of the function well-managed, I split the implementation into a header and a .c source file.

After this change, the benchmarked runtime improved slightly from 35.1 sec to 33.5 sec.

ngoldbaum reviewed (commented):
I think the implementation of fnv-1a in this PR isn't correct. Maybe we should just be using the (public-domain licensed) reference implementation: https://github.com/lcn2/fnv.

I didn't look closely at the rest after I noticed this issue.

math-hiyoko committed:
fix: cast using const void *

math-hiyoko committed:
fix: fix fnv1a hash

math-hiyoko committed:
fix: lint

math-hiyoko committed:
35.1sec -> 33.5sec

ngoldbaum commented:
Sorry this didn't quite make it in time for NumPy 2.3, but i'm going to try to get this merged ASAP so we have time to iterate in-tree as needed. I'll try to give this another once-over sometime this week.

I also want to say that this is a really impressive contribution and any delay on my part is because I have to focus on other stuff and it takes a lot of time and attention to properly review CPython C API code rather than me being unenthusiastic about getting this improvement into NumPy.

math-hiyoko committed:
Merge branch 'main' into feature/#28364

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +24,243 @@ struct FinalAction {
+    // - Reserving for all elements (isize) may over-allocate when there are few unique values.
+    // - Using a moderate upper bound (1024) keeps memory usage reasonable (4 KiB for pointers).
+    // See discussion: https://github.com/numpy/numpy/pull/28767#discussion_r2064267631
+    std::unordered_set<T> hashset(std::min(isize, (npy_intp)1024));
```
can you make this an ALL_CAPS_MACRO defined at global scope instead of the literal 1024?

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +24,243 @@ struct FinalAction {
+    char *idata = PyArray_BYTES(self);
+    npy_intp istride = PyArray_STRIDES(self)[0];
+    for (npy_intp i = 0; i < isize; i++, idata += istride) {
+        hashset.insert(*(T *)idata);
```
According to [cppreference](https://www.cppreference.com/w/cpp/container/unordered_set/insert.html), this overload of `insert` returns:

>A pair consisting of an iterator to the inserted element (or to the element that prevented the insertion) and a bool value set to true if and only if the insertion took place.

So I think you need to handle the case where the insertion doesn't happen.

math-hiyoko replied:
In this case, we're only collecting unique values, so it's actually fine if an insertion doesn't happen — duplicates are expected and safely ignored.
Is there something specific you'd suggest checking or handling in that case?

ngoldbaum replied:
I was thinking of allocation failures, but that's probably not worth worrying about on second thought

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +24,243 @@ struct FinalAction {
+    char *idata = PyArray_BYTES(self);
+    npy_intp istride = PyArray_STRIDES(self)[0];
+    for (npy_intp i = 0; i < isize; i++, idata += istride) {
+        hashset.insert((T *)idata);
```
need to check return value and handle possible failure

ngoldbaum reviewed (approved):
I see some minor issues that should be addressed but this is ready to merge then. I looked all the C++ code over in detail and didn't spot any issues besides these.

ngoldbaum commented:
Also you might be interested in https://github.com/numpy/numpy/issues/29229

math-hiyoko committed:
enh: define macro HASH_TABLE_INITIAL_BUCKETS

math-hiyoko committed:
enh: error handling of NpyString_load

math-hiyoko committed:
enh: delete comments on GIL

math-hiyoko commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -18,77 +24,243 @@ struct FinalAction {
+    NPY_ALLOW_C_API_DEF;
+    NPY_ALLOW_C_API;
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    // NumPy API calls and Python object manipulations require holding the GIL.
```
I've removed the GIL-related comments as suggested.

math-hiyoko commented:
Thanks so much for the thorough review! This was my first PR to NumPy, and your detailed feedback really helped me bring it to completion.

I'll take a look at #29229 as you suggested — I'm also interested in #28363, which I think I can complete in a similar way to this PR.

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -249,7 +249,11 @@ unique_vstring(PyArrayObject *self, npy_bool equal_nan)
-        NpyString_load(in_allocator, packed_string, &unpacked_strings[i]);
+        int is_null = NpyString_load(in_allocator, packed_string, &unpacked_strings[i]);
+        if (is_null == -1) {
+            // failed to load string
```
I think you need to explicitly set an error indicator with e.g. `npy_gil_error`. `NpyString_load` doesn't do it for you.

math-hiyoko committed:
fix: PyErr_SetString when NpyString_load failed

ngoldbaum commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -243,7 +243,8 @@ unique_vstring(PyArrayObject *self, npy_bool equal_nan)
         int is_null = NpyString_load(in_allocator, packed_string, &unpacked_strings[i]);
         if (is_null == -1) {
-            // failed to load string
+            PyErr_SetString(PyExc_RuntimeError,
```
you need to have the GIL acquired before you call this - that’s why I suggested `npy_gil_error`, since it handles that dance

math-hiyoko replied:
Thank you for pointing this out.
Switched `PyErr_SetString` to `npy_gil_error`.

math-hiyoko committed:
fix: PyErr_SetString -> npy_gil_error
